### DISCUSSION TITLE
Great Job. Quick Question re; multiple Servo's

### DISCUSSION THREAD

**Discussion by TimRosener** [joshua-8/Derivs_Limiter#19] [Q&A] [answered]
How do you recommend handling multiple servos? I have a project with 16 servos with different PWM limits and velocity profiles for smoothing.

**Comment by joshua-8** [comment, marked as answer]
Hi @TimRosener,
Thank you for your interest in this library!

You can make as many instances of Derivs_Limiter as you want. You can have one Derivs_Limiter per servo and have separate velocity and range limits for each servo. This library doesn't help with actually sending control signals to many servos, it only provides the smoothed angle to send to the servo.

Do you have independent control of all 16 of your servos working? (without velocity profiles)
Controlling 16 servos is a challenge for a standard Arduino. The standard library has a limit of 12 servos https://www.arduino.cc/reference/en/libraries/servo/
If you haven't already found a way to control 16 servos, here are some suggestions:
- There might be a library that supports more servos.
- While I have no experience with this, if you search for `arduino servo driver board` there are results for boards that can connect to your Arduino and provide more servo outputs.



**Thanks to you, I realized that I didn't have an easy way to handle an array of Derivs_Limiter objects.**
**I made a fix for that, and added [an example](https://github.com/joshua-8/Derivs_Limiter/blob/master/examples/Multi/Multi.ino) for using multiple copies of Derivs_Limiter.**

**Please update the library to 3.1.0 or newer and let me know how it works for you.**

For some reason the new library version isn't showing up in Arduino's library manager, so you may have to download the updated library from here: https://github.com/joshua-8/Derivs_Limiter/releases/tag/v3.1.1

Here is an image showing 4 independent Derivs_Limiters running at once (it's the output of the Multi example).
![image](https://user-images.githubusercontent.com/59814881/202335983-2485d369-d9d3-4f4e-9e89-5f7ff5b4eaa4.png)

It sounds like you're working on a very cool project! Good luck, and feel free to ask any other questions you might have.

-- joshua-8

**Reply by TimRosener** [reply]
This is great! Thank you! 

I am running this on a Rpi Pico W. 

There is a [library ](https://github.com/khoih-prog/RP2040_ISR_Servo)that was created with the RP2040 dual-core microcontroller in mind. The RP2040 has 8 state machines that can drive 2 PWM signals each without tying up the dual-core processors. Still working through the details; however, I have Servo's running. I plan on having one core manage the servos and the second core manages a UI via Wifi for the configuration and tuning of servos profiles. All of this is for driving animatronics via DMX show control protocols. 

I am going to try the 3.1 version you created; thanks for doing that. That seems to be the cleanest and most efficient way to implement. 

Quick Question, one drawback of the RP2040 is its lack of hardware-based floating point operations. How intensive are the FP operations in your implementation? Will I be pushing things to have 16 instances of the servo profiles running at the same time?  I have some flexibility in design, meaning I could move the DMX receive code block to the second core.

Thanks again for a great library and lighting-fast responses to my questions. 

Cheers.

**Reply by joshua-8** [reply]
The RP2040 sounds cool!
It's great to hear you have the servos running!

The library uses a lot of floating point operations; it is not optimized for speed in any way.

I just ran a test with an Arduino Uno and it was able to calculate 16 motion profiles 93 times per second. That should be fast enough for the servos to look smooth. The Uno doesn't have an FPU and only runs at 16Mhz, so since you're using the RP2040 which has a clock speed several times faster, I would expect to be fine. Microcontrollers are pretty fast!

Cheers!

**Reply by TimRosener** [reply]
That should be more than fast enough. Thank you!

**Reply by joshua-8** [reply]
Feel free to ask any other questions as you work on your project. I'm really happy to hear from someone using my library!

**Reply by TimRosener** [reply]
Quick Update, ran some tests. Performance is not an issue; in fact I had to use some timers to slow down the serial output so that I could see the waveforms. Thanks again.

**Reply by TimRosener** [reply]
<img width="1214" alt="screenshot_191" src="https://user-images.githubusercontent.com/179073/202539068-097d2c89-d605-446b-85f9-ce510403f8a0.png">

**Reply by joshua-8** [reply]
I'm happy to hear that!

**Reply by TimRosener** [reply]
Another quick question, your documents talk about velocity in terms of intervals per second; I assume that the Acceleration parameters are intervals/Sec Squared.  working with hobby servo, I have ranges of 0-180 or 0-270 degrees that I am operating in.  can you give more insight on how you calculate movement per second? I have had to increase values by a factor of 50 to get some of the profiles I need, for instance, a value of 12,500 (256 DMX Channel max range * 50) to get the desired response to match the input response. A value of 250 gives me a very slow response compared to the input position. If we were working with degrees (Intervals) per second, it would seem my working ranges would be more in the 0-200 range for velocity and acceleration parameters. 

One of the reasons I am asking is that I want to calculate the right max parameters based on the Servo specs; many times, those values are given in the time it takes a servo to move 60 degrees. Then allow the velocity and acceleration limits to operate within those ranges. BTW I am dynamically adjusting those acceleration parameters on the fly. 

Thanks again for this great library!

**Reply by joshua-8** [reply]
If you are controlling the servos by sending an angle in degrees, then I think DL.getPosition() is degrees, DL.getVelocity() is degrees per second and DL.getAcceleration() has units of degrees per second per second.

You might need to use a small fraction of a servo's max rated speed to make sure that the servo moves smoothly, since otherwise the servo might lag behind the position that you're trying to set it to.

A value of 12500 for acceleration or velocity sounds really large to me, I'm not sure why you're needing that. **Are you able to make a graph of position, velocity, and acceleration?**

As a point of reference, here is a video of a servo moving between 0 and 180 degrees, with a maximum velocity of 500 degrees per second and an acceleration of 2000 degrees per second per second.

https://user-images.githubusercontent.com/59814881/202831349-42a95536-056f-43d0-a15b-9285732fc465.mp4

**Reply by TimRosener** [reply]
Thanks for looking into this. I took a step back and went through some of my code. I was able to get it working as you described. I had an error when I was dealing with acceleration. Your note helped a lot. Thank you. 

You had asked about getting 16 servos working on the RP2040. The only way I could get that to work was by taking advantage of the State Machines. RP2040 has 8 of them each can control twp PWM pins. There is a great [library](https://github.com/khoih-prog/RP2040_PWM) that allows you to drive the state machines. This has the added advantage of freeing up the core from all of the PWM management. I just had to build a servo control layer on top of it. 

Thanks again, cheers and happy holidays!

**Reply by joshua-8** [reply]
I'm happy to hear that it's working for you now.
Thanks for pointing out the library for PWM state machines!
Best of luck with all your projects.
Cheers and happy holidays!
